### PR TITLE
Adding support for TestNG's preserveOrder and groupByInstances options

### PR DESCRIPTION
**Pull request by riemenschneider**
Currently Gradle does not support TestNG's preserveOrder and groupByInstances options. These options are required if you want to enforce that tests are run completely before the next test is processed. (Otherwise you might end up with 'execution plans' like (Test1.@BeforeClass, Test2.@BeforeClass, Test1.@Test, Test2.@Test) which causes problems if both @BeforeClass methods try to allocate the same unique resource.)

I already discussed this enhancement [in the gradle-dev group] (https://groups.google.com/d/topic/gradle-dev/gQmLHgtYJAk/discussion). Since there were no arguments against supporting these options I have prepared an according pull request.

Any feedback is appreciated!

### CODE REVIEW

**Comment by breskeby** [comment]
Hey the pull-request looks good, but would you mind adding some integration test coverage. Have a look at https://github.com/gradle/gradle/tree/master/subprojects/plugins/src/integTest/groovy/org/gradle/testing/testng for examples of testng related integration tests

**Comment by riemenschneider** [comment]
I added some integration tests. Unfortunately two of them are blocked by https://github.com/cbeust/testng/issues/639. The issue is already fixed in TestNG's Git repository but not yet released.

**Comment by riemenschneider** [comment]
The next TestNG release is planned for next week (https://groups.google.com/forum/#!topic/testng-dev/e4D1gCeMQZ0). I will retest the integration tests using the new release and update if necessary.

**Comment by riemenschneider** [comment]
TestNG 6.9.4 containing the fix for https://github.com/cbeust/testng/issues/639 is out since May 9th. I have updated TestNGCoverage.NEWEST and now all integration tests are running as expected.

**Comment by riemenschneider** [comment]
The requested integration tests are available for quite a while. Any reason for not merging the pull request?

**Comment by stefanhendriks** [comment]
+1, really wanting this feature :+1:

**Comment by riemenschneider** [comment]
Same for me...

Waiting for the integration of this pull request since I need it for our product's build process (I am currently stuck with a rather old TestNG version).

Software AG – Sitz/Registered office: Uhlandstraße 12, 64297 Darmstadt, Germany – Registergericht/Commercial register: Darmstadt HRB 1562 - Vorstand/Management Board: Karl-Heinz Streibich (Vorsitzender/Chairman), Eric Duffaut, Dr. Wolfram Jost, Arnd Zinnhardt; - Aufsichtsratsvorsitzender/Chairman of the Supervisory Board: Dr. Andreas Bereczky - http://www.softwareag.com

**Comment by pioterj** [comment]
The Gradle team doesn't currently have the capacity to deal with all pull requests in a timely fashion. We expect this situation to be temporary. We ask for patience in the meantime.

**Comment by pioterj** [comment]
Actually, since this is a simple change I'll be able to merge this sooner than I thought. @riemenschneider Could you resolve the conflicts first?

**Comment by riemenschneider** [comment]
I will do so.

**Comment by riemenschneider** [comment]
Conflicts are resolved.

**Comment by pioterj** [comment]
@riemenschneider Thank you for the work you did so far. I took a closer look at this pull request and here are my comments:

- Annotate new fields in TestNGOptions with @Incubating. This is what we always do for new stuff in the API. 
- Use primitive boolean type for new fields in TestNGOptions. This is what we normally do in other places and it will make the code a little bit safer and simpler. Let's default to what was previously used for backward compatibility (both values false when not explicitly set to true)
- One integration test case per new option should be perfectly enough. These tests are very slow, especially when executed for multiple versions. You could try to combine similar test cases and rely on unit tests to verify default values and corner cases.
- Last two test cases for preserveOrder and groupByInstances are almost identical. It would be good to use specific [minimal fixture](http://xunitpatterns.com/Minimal%20Fixture.html) so that both tests clearly document what each option is about. If this comment sounds too vague please let me know and I'll be happy to help you to come up with better test cases.
- Remove TextUtil.toPlatformLineSeparators when checking output in the integration test. This causes tests to fail on Windows because output is normalized to use Unix style newlines.
- Update user guide under "23.13". A new section called something like "Controlling test execution order in TestNG" should do the trick.

**Comment by riemenschneider** [comment]
Feedback is incorporated.

**Comment by mark-vieira** [comment]
Thanks @riemenschneider. We are currently preparing for the release of Gradle 2.10. We should be able to merge this soon for inclusion in Gradle 2.11.

**Comment by pioterj** [comment]
@riemenschneider Thank you. There are two major problems with the corrections you made:

- New integration tests fail when executed against multiple TestNG versions. Note that the "org.gradle.integtest.multiversion" property needs to be set to "all" in order to run tests against multiple versions
- Documentation can't be generated. You can verify that using command "gradlew :docs:docs"

Also, I noticed that you made a comment in the user guide that Gradle's overrides TestNG defaults in regards to preserving test execution order. This is not our intention. The documentation says that tests run in the order they are found in the XML file by default (when preserve-order attribute is not set). However when I checked org.testng.TestNG class which Gradle uses to run tests it explicitly sets this attribute to false. In case the defaults vary between TestNG versions we should reconsider how these new values are set. It would be great if you could do some research into that. In general we want to avoid introducing changes that are not backward compatible. If someone uses specific TestNG version with Gradle now we don't want test execution order to change when when they upgrade Gradle.

**Comment by pioterj** [comment]
As for dealing with features that are not supported in specific versions, you can take a look at tests for JUnit. A good way to do it would be to add a new constant in TestNGCoverage and have separate tests for versions that support and another one in a version that does not support it.

**Comment by riemenschneider** [comment]
I have split up the integration into two separate integration tests for preserve-order and group-by-instances and created separate coverages for them. Unfortunately the TestNG version supporting preserve-order without issues is 6.9.4 and TestNGCoverage defines the NEWEST constant based on the Java 7 compatibility of the JVM. There I did the same trick for the preserve-order coverage which is ['6.9.4', NEWEST] if the JVM supports Java 7 and empty otherwise.

Do we have any common handling for the situation that an option is used with a TestNG version not supporting it? Currently it is silently ignored.

The documentation issue is also fixed.

**Comment by pioterj** [comment]
Thanks, I'll take a closer look at updated changes and reply shortly.

File: subprojects/docs/src/docs/userguide/javaPlugin.xml [hunk 1]
```diff
@@ -1349,6 +1349,26 @@
                 <sourcefile file="build.gradle" snippet="test-config"/>
             </sample>
         </section>
+
+        <section id="test_execution_order">
+            <title>Test execution order</title>
+            <para>TestNG allows explicite control of the execution order of tests.</para>
```
Comment (paulk-asert) [inline subprojects/docs/src/docs/userguide/javaPlugin.xml]: minor typo: explicite should be explicit

**Comment by pioterj** [comment]
Thank you for your persistence in implementing all the changes. We are really close to get this merged. I added few more comments to the latest version.

I think it would be least surprising for the user if we fail early in case an attempt is made to use an option that is not supported in specific TestNG version. But I see in case of setAnnotations we do ignore such case silently. In either case we'll need a test case for that. @mark-vieira Do we have any established practice to deal with such cases?

**Comment by mark-vieira** [comment]
In most cases we want to go with the default behavior of the framework we are integrating with. If TestNG aims for forward compatibility and ignores unknown options then we should do the same. In most cases I say just pass the options along and let TestNG deal with it and propagate the failure (if any).

**Comment by riemenschneider** [comment]
What is the default behaviour of TestNG from your point of view?

On the one hand you can configure the test suite programmatically by using org.testng.TestNG API. Using that way a NoSuchMethodException is more or less the way TestNG is handling the situation (the method is not there and if you try to call it using reflection, you get a NSME).

On the other hand TestNG is mostly configured using a testng.xml file and if this XML configuration contains an unknown attribute, it is simply (and silently) ignored.

I have currently implemented the first way (try to set the option and if that fails AND the option is true, throw the caught exception) in order to inform the user about the fact, that is was not possible to run the test suite as it was specified. Unfortunately I am not able to pass the option to TestNG and propagate failures since there are not methods for passing the options.

**Comment by mark-vieira** [comment]
Understood, wasn't sure if the TestNG API was static or if options were passing in as a map of string or something similar. In this case we would have to handle the NSME, the question is whether we fail the build or not. I don't have super strong feelings on this either way but in general I'm not a huge fan of silently ignoring errors. I think returning an error indicating to the user that the option isn't available on the version of TestNG they are using seems reasonable. I would make sure we include this information in the documentation (JavaDocs, DSL reference, etc), indicating that these options are only applicable to certain versions of TestNG.

**Comment by mark-vieira** [comment]
Of course this also means test coverage for that above scenario :)

**Comment by pioterj** [comment]
@riemenschneider It would be perfect if you could also update the design document at https://github.com/gradle/gradle/blob/master/design-docs/testing-improvements.md with these changes. Normally we do it before an implementation starts. This time it will be post-factum but still a valuable exercise if you are interested in contributing to Gradle more often.

**Comment by riemenschneider** [comment]
I modified the error handling for the case that one of the options is set to true AND TestNG does not support the according feature and added notes to the documentation. Integration tests for the error handling are now also in place.

Additionally  I updated the design document.

**Comment by pioterj** [comment]
Great! I'll merge it next week.

**Comment by riemenschneider** [comment]
I see two SpockTimeoutErrors that I am not able to reproduce in my environment.

**Comment by pioterj** [comment]
I'll try to restart the build. Maybe it was some random failure.

**Comment by pioterj** [comment]
It's merged. Thank you for the contribution. It will be available in Gradle 2.11.
